### DISCUSSION TITLE
🧮 publicodes vs tableur de calcul basique

### DISCUSSION THREAD

**Discussion by laem** [publicodes/publicodes#88] [Ideas]
Chez nosgestesclimat nous sommes confrontés à un "problème" assez intéressant, je vous le décris ici. @clemog

Assez intéressant, car on s'est rendus compte que publicodes peut se voir comme un remplaçant des outils de tableur dans beaucoup de cas, mais qu'on est ici confrontés à une limite importante. Intéressant, car ça ouvre le besoin à un système de paquets à la NPM.

## Une opération sur un tableau comme valeur de variable 

Notre objectif est d'intégrer dans nos bases de règles le résultat d'une somme sur un simple tableau à plusieurs colonnes (les attributs des items) et plusieurs lignes (les instances de ce type d'objet). 

Parlons concret : aujourd'hui nous avons ça dans notre base de code. 

```yaml

alimentation . plats . végétarien:
  formule: empreinte * nombre
  icônes: 🥗🍳🧀
  description: |
    Repas sans chair animale, avec des oeufs ou un peu de fromage.
    Voici le menu type utilisé pour notre calcul.
    | Ingrédients  | Qte (g)  | Kcal  | Protéines (g)  | Lipides (g)  | gCO2e /kg  | gCO2e   |
    |:-:|:-:|:-:|:-:|:-:|:-:|:-:|
    |      Entrée : soupe de légumes  |  |  |  |  |  |  |
    | 200g légumes de saison  | 200  | 60  | 3  | 0  | 267  | 53.4  |
    | huile (1/2 c.s.)  | 7  | 63  | 0  | 7  | 2153  | 15,1  |
    |      Plat principal : omelette aux pommes de terre et aux oignons  |  |  |  |  |  |  |
    | 2 œufs  | 106  | 195  | 16  | 11  | 2610  | 276,7  |
    | pommes de terre  | 200  | 170  | 3  | 0,2  | 77  | 15.5  |
    | huile (1/2 c.s.)  | 7  | 63  | 0  | 7  | 2153  | 17,1  |
    |      Dessert : salade de fruits  |  |  |  |  |  |  |
    | fruits de saison  | 200  | 100  | 2  | 0  | 267  | 53.4  |
    | Pain  | 50  | 125  | 4  | 0,6  | 1520  | 76  |
    | Total  |    | 776  | 28  | 25,8  |    | 510 |

```
https://github.com/datagir/nosgestesclimat/blob/master/data/alimentation.yaml#

On a ensuite une autre variable qui représente la somme de la colonne `gCO2e` : 

```yaml

alimentation . plats . végétarien . empreinte:
  formule: 0.510
  unité: kgCO2e
  note: C'est le "Repas végétarien 1" [ici](https://www.bilans-ges.ademe.fr/documentation/UPLOAD_DOC_FR/index.htm?repas.htm).
```

Pour ceux qui ne connaissent pas, la syntaxe du tableau dans la `description` n'est pas inventée, c'est ce qui permet d'afficher un beau tableau dans la doc [ici](https://nosgestesclimat.fr/documentation/alimentation/plats/végétalien). 

![Capture d’écran du 2021-10-25 18-28-55](https://user-images.githubusercontent.com/1177762/138734451-84ad56c7-4451-45da-acc9-e36a1e4ef69e.png)

L'implé actuelle pose plusieurs problèmes : 
- le calcul est fait à la main, avec le risque d'erreur très élevé associé, car le tableau n'est que figuratif, pas la base de la valeur  de `empreinte` 
- il faut utiliser un utilitaire de conversion de tableur (CTRL-C dans libreoffice) vers le tableau markdown pour documenter la variable avec ces informations tout à fait essentielles à la compréhension du calcul

Tout le monde conviendra que l'affichage du tableau en markdown, malgré ces problèmes, est bien plus cool que si l'on avait réimplémenté le calcul avec la syntaxe actuelle publicodes, qui aurait donné un truc comme ça : 

```yaml
repas végé . huile: 
  description: (1/2 c.s.) 
repas végé . huile . quantité: 7g
repas végé . huile . calories : 63 kCal
#... (lignes)
repas végé . huile . empreinte: 15,1 gCO2e

#... (colonnes)

repas végé . [prochain ingrédient] : 
```

On y aurait passé la journée, pour un résultat moins satisfaisant qu'un tableur (même si ici l'avantage est qu'on peut distinguer le nom de l'attribut (quantité) et l'unité (grammes)).

Le problème est évident : impossible de représenter un format à deux dimensions dans un éditeur de texte ! On jongle alors avec des inconvénients : soit c'est verbeux (syntaxe objet style JSON) et on répète chaque attribut n fois, soit c'est peux verbeux (CSV) mais les colonnes ne sont pas alignées et on comprend donc rien. 

## Inception de tableaux 

Mais notre problème va encore plus loin : nous allons maintenant jusqu'à intégrer des ingrédients "proxy" qui représentent une moyenne d'ingrédients. 

Concrètement pour intégrer le menu 🐟️ poisson, on va faire la **moyenne** et non plus la somme de l'empreinte de 10 poissons différents, puis injecter cette moyenne dans une colonne d'une ligne du tableau décrit précédemment, pour recalculer l'empreinte totale du menu. 

![Capture d’écran du 2021-10-25 18-44-15](https://user-images.githubusercontent.com/1177762/138736403-fa903534-c345-4d2c-ae03-5c6fbbbfa3d1.png)

Il est alors encore plus évident que le format de travail sur ce genre de données est bien le tableur. C'est l'archétype du tableur faiblement calculatoire, mais riche en données.

Le plus logique me semble donc de réserver publicodes au cas d'usage "inverse" : des tableurs avec peu de données mais fortement calculatoires, avec des formules qui se référencent de partout. 

## Inflation d'un dépot de règles

J'aborde un dernier problème : notre modèle de calcul de l'empreinte du poisson pourrait être complètement séparé du modèle de calcul général de nosgestesclimat, qui rassemble plein d'autres sujets n'ayant rien à voir. J'en fais une autre discussion, #89, pour ne pas polluer celle-ci :) 

## Solutions possibles

Ce qu'on prévoit pour l'instant c'est simplement de **référencer des tableurs mise en ligne comme des fichiers suivis sur github**. Le problème, outre la rupture entre nosgestesclimat.fr/documentation et l'ouverture d'un fichier tableur, ainsi que l'écueil d'un copier coller nécessaire et source d'erreur ("OUPS j'ai mis à jour le tableur mais pas mis à jour la valeur dans la base de règles") c'est que le suivi des diff sur github est inexistant. Il semble exister des outils pour ça comme [xltrail](https://www.xltrail.com), mais je doute que ce soit pratique.

Autre possibilité : permettre lors de la compilation d'une base de règle la lecture de cases dans un tableur. Ça nécessiterait d'importer du code qui lis un fichier pour en extraire une valeur calculée, mais l'avantage serait un rapprochement du monde des tableurs, avec potentiellement à la clef la conversion de règles publicodes en tableur ? En tout cas un outil comme [sheetJS](https://github.com/sheetjs/sheetjs) permet de le faire et semble super fiable. 

![Capture d’écran du 2021-10-25 19-23-39](https://user-images.githubusercontent.com/1177762/138741644-12b3f567-8af8-4732-a7d1-133ec5b9377b.png)
![Capture d’écran du 2021-10-25 19-24-57](https://user-images.githubusercontent.com/1177762/138741795-26e98bc2-5bbf-4d85-95a9-8f182bc34d30.png)

⬆️ aperçu du contenu du fichier content.xml d'un fichier .ods ultra simple qui n'est autre qu'un dossier zippé de contenu texte

On pourrait aussi définir un nouveau mécanisme qui se chargerait de pouvoir faire des sommes sur des tableaux pas trop grands, qui "tiennent" dans une base publicodes. Voir #26 qui est très lié. On peut même imaginer faire références à des fichiers YAML "purs" pour ces petits tableaux. 

Autre solution qui me vient à l'esprit en écrivant tout ça, c'est de faire référence à des fichiers CSV qui côtoieraient les .yaml, pour les gros tableaux. Avantage : non seulement les fichiers CSV sont visualisables sur github, mais le diff aussi !  Les logiciels de tableau (LibreOffice, Excel, Google docs) ouvrent nativement des CSV pour travailler et mettre à jour ces données.

![Capture d’écran du 2021-10-25 19-10-55](https://user-images.githubusercontent.com/1177762/138740029-d4d3764e-935a-4eb4-a8dc-24db49b38f65.png)

On pourrait alors charger un tel CSV et définir des calculs dans publicodes. On pourrait alors reproduire des modèles de tableurs assez complexes, avec l'immense avantage d'être suivis sur github. Cependant, pour de gros CSV ça demande de charger en mémoire, sans lazy loading, des données qui pourraient réduites à un seul chiffre, et ça éloigne les experts métiers qui préféreraient définir des formules de calcul dans un tableur. 

Une dernière idée qui me semble très intéressante : le tableur serait hébergé en ligne (on pense à Google docs, mais il faudrait pouvoir reposer sur du libre, par exemple [collabora](https://www.collaboraoffice.com/fr/collabora-online/)) et exposerait une API qui permettrait de récupérer la valeur calculée d'une case. Je pense que [Gdoc le fait](https://developers.google.com/sheets/api/guides/values). Le désavantage, c'est qu'on aurait alors une autre infrastructure que github pour gérer une partie du code, avec des droits d'éditions à gérer, et une dépendance à un service qui peut tomber.

> Note : il existe plein de moyens pour visualiser dans la doc publicodes, en lazy loading, le contenu d'un tableur. [exemple](https://github.com/mengshukeji/Luckysheet). 

Voilà, gardons en tête que c'est un besoin pas urgent, donc à vos idées !

**Comment by mquandalle** [comment]
Merci pour les exemples super intéressants. Effectivement référencer un csv (que l'on pourra éditer dans Libreoffice) semble être la solution la plus simple. Ça impose de garder les formules de calcul dans un fichier publicodes plutôt que dans un tableur associé, ce qui facilite grandement l'implémentation. On aurait seulement de fonctions d'agrégation “somme”, “moyenne”, etc. un peu comme dans Airtable par exemple : 
![image](https://user-images.githubusercontent.com/1730702/138842740-ebe59075-a5a9-4d8c-9588-d1120c593a5f.png)

Pour l'idée d’héberger le tableur en ligne, ça doit pouvoir fonctionner si l'éditeur utilisé expose un fichier csv sur une URL type https://collaboraoffice.com/ademe/mix-eletrique.csv

Autre point on avait discuté oralement de la possibilité de ne pas télécharger tout le publicodes source `impact-poissons.publicodes` mais seulement un aggrégat comme une moyenne. Pour faire cela, je pense c'est au paquet source de publier directement les valeurs agrégées par exemple dans un fichier `moyennes.json` généré à la publication du paquet.

**Reply by mquandalle** [reply]
Google Sheets par exemple permet de récupérer un csv https://stackoverflow.com/a/33727897

**Reply by mquandalle** [reply]
À noter aussi que si on introduit des fonctions d’agrégation, on pourrait s'en service pour faire une somme au niveau d'un espace de nom, sans avoir à référencer chacun des enfants (et risquer d'en oublier). ex: somme des cotisations, somme des impacts co2, etc.

**Reply by lajarre** [reply]
Csv sans calculs: top. Mais il faut garder un versioning sur les csvs, n'est-ce pas? Pas sûr que les gsheets et autres puissent donner totale satisfaction sur ce point.

**Reply by laem** [reply]
Oui, pour versionner les CSV, ça me semble idéal de les mettre sur le git près des .yaml. 

Par contre, ça limite alors ces tableurs à des tableaux simples, sans calculs. Pour une 1ère version, ça me semble déjà vachement bien. En plus, ça permet de bien séparer les traitements en plusieurs bouts autonomes, assemblables en publicodes.
